pymongo/mongo_client.py

```diff
@@ -1635,7 +1635,7 @@ def _cleanup_cursor(
             # The cursor will be closed later in a different session.
             if cursor_id or conn_mgr:
                 self._close_cursor_soon(cursor_id, address, conn_mgr)
-        if session and not explicit_session:
+        if session and not explicit_session and get_thread_client_session() is None:
```

Comment:
Can't we let the `_end_session` call take care of it?

Reply:
I think putting it above end session is less risky.